EIPS/eip-5089.md

```diff
@@ -0,0 +1,340 @@
+_Alberto: Optionality is evil._
+
+MUST emit the `Redeem` event.
+
+MUST support a redeem flow where the Principal Tokens are burned from `owner` directly where `owner` is `msg.sender` or `msg.sender` has ERC-20 approval over the shares of `owner`.
+MAY support an additional flow in which the shares are transferred to the Principal Token contract before the `redeem` execution, and are accounted for during `redeem`.
+
+MUST revert if all of `amount` cannot be redeemed (due to withdrawal limit being reached, slippage, the owner not having enough Principal Tokens, etc).
+
+Note that some implementations will require pre-requesting to the Principal Token before a withdrawal may be performed. Those methods should be performed separately.
+
+```yaml
+- name: redeem
+  type: function
+  stateMutability: nonpayable
+
+  inputs:
+    - name: amount
+      type: uint256
+    - name: to
+      type: address
+    - name: from
+      type: address
+
+  outputs:
+    - name: assets
+      type: uint256
+```
+
+#### mature
+_Alberto: We surely don't need this method in the standard. It is up to each PT creator to decide how they mature it._
+
+Callable once `block.timestamp >= maturity`, stores the current `exchangeRate` as the `maturityRate` and "matures" the Principal Tokens. 
+
+OPTIONAL - This method can be used to provide needed data for the calculation of interest post-maturity, but interfaces and other contracts MUST NOT expect this method to be present.
+
+MUST emit the `Mature` event.
+
+MUST revert if `block.timestamp < maturity`.
+
+```yaml
+- name: mature
+  type: function
+  stateMutability: nonpayable
+
+  inputs: []
+
+  outputs:
+    - name: exchangeRate
+      type: uint256
+```
+
+### Events
+
+#### Mint
+
+`from` has exchanged `amount` of underlying for `amount` of Principal Tokens, and transferred those Principal Tokens to `to`.
+
+MUST be emitted when underlying is deposited and Principal Tokens are minted from the contract in the `ERC5070.mint` method.
+
+```yaml
+- name: Mint
+  type: event
+
+  inputs:
+    - name: from
+      indexed: true
+      type: address
+    - name: to
+      indexed: true
+      type: address
+    - name: amount
+      indexed: false
+      type: uint256
+```
+
+#### Redeem
+
+`from` has exchanged `amount` of Principal Tokens for `amount` of underlying, and transferred that underlying to `to`.
+
+MUST be emitted when Principal Tokens are burnt and underlying is withdrawn from the contract in the `ERC5070.redeem` method.
+
+```yaml
+- name: Redeem
+  type: event
+
+  inputs:
+    - name: from
+      indexed: true
+      type: address
+    - name: to
+      indexed: true
+      type: address
+    - name: amount
+      indexed: false
+      type: uint256
+```
+
+#### Mature
+_Alberto: Maybe we need this event, even if we don't have a `mature` function._
```

Comment:
I think we do.

In many cases `maturity` doesnt actually mean the market has matured with reference to further yield generation and where it gets directed.

Reply:
Yeah, agreed.